dev-docs/RFCs/v6.2/mapbox-custom-layer-rfc.md

```diff
@@ -0,0 +1,191 @@
+# RFC: Mapbox Custom Layer Integration
+
+* Authors: Xiaoji Chen, Ib Green
+* Date: Sep 25, 2018
+* Status: **Draft**
+
+References:
+
+* [Layer Group Operations RFC](v6.x/layer-and-group-operation-rfc.md)
+
+
+## Abstract
+
+This RFC proposes changes to the deck.gl API to allow deck.gl to use mapbox context and to trigger rendering of certain deck.gl layer renders in between mapbox style layers.
+
+
+## Problems
+
+* Initializing the DeckGL instance with an external gl context. (Work towards making this as declarative as possible).
+* Render deck.gl "layer-by-layer" - Work towards an official "static" rendering API for deck.gl
+
+
+## Proposed MapboxGL API Integration
+
+This case is fairly simple: a `MapboxLayer` class that can be used as a custom mapbox layer provides the plumbing and enables mapbox users to use deck.gl "layer packs".
+
+The adapter class transparently creates and manages as `Deck` instance under the hood.
+
+```js
+import Deck, {ScatterplotLayer} from '@deck.gl/core';
+import {MapboxLayer} from '@deck.gl/mapbox';
+
+const map = new mapboxgl.Map({
+  container: 'map',
+  style: 'mapbox://styles/mapbox/light-v9',
+  center: [-74.012, 40.705],
+  zoom: 15.5,
+  bearing: -20,
+  pitch: 45
+});
+
+map.on('load', () => {
+  map.addLayer(mapboxBuildingLayer);
+  map.addLayer(deckPoiLayer, getFirstTextLayerId(map.getStyle()));
+  map.addLayer(deckRouteLayer);
+});
+```
+
+
+## Proposed DeckGL API Integration
+
+This case is a bit harder since it both requires postponing deck.gl initialization.
+
+### Proposal 1: Callbacks
+
+```js
+import Deck, {ScatterplotLayer} from '@deck.gl/core';
+import {MapboxLayer} from '@deck.gl/mapbox';
+
+const map = new mapboxgl.Map({...});
+
+const deck = new Deck({
+  width: '100%',
+  height: '100%',
+  longitude: -122.4,
+  latitude: 37.8,
+  layers: [
+    new MapboxLayer({
+      id: 'my-scatterplot',
+      insertBefore: 'admin-1-labels',
+      type: ScatterplotLayer,
+      ...
+    })
+  ],
+  onAddLayer: (layer) => {
+    if (layer instanceof MapboxLayer) {
+      map.addLayer(layer, layer.props.insertBefore)
+    }
+  },
+  onRemoveLayer: (layer) => {
+    if (layer instanceof MapboxLayer) {
+      map.removeLayer(layer.id);
+    }
+  }
+});
+```
+
+Advantages:
+
+* Allows deck.gl and mapbox layers to be ordered (to some degree) within the deck.gl layer list
+* Callbacks are generic and could possibly be useful in other contexts
+
+Downsides:
+
+* Uses `MapboxLayer` inside deck.gl API, ideally that module should be reserved for mapbox API.
+* Relies on callbacks which is not "declarative"
```

Comment:
The reason custom callbacks are used here is that the core `Deck` class does not know about mapbox. The path to supporting JSON layers will be similar to how we support the base map in the JSON API, where we eliminate the need for `onViewStateChange`.

Reply:
> the path to supporting JSON layers will be similar to how we support the base map in the JSON API

I am not sure I understand your points here. Are we talking about mapbox layers that just take JSON props?

If not (i.e. the intention was "the path to supporting mapbox layers will be..."), maybe switch the sample code for proposal 1 to a React API example to help clarify this. I think that is the more important / slightly harder case.